File: auxil/process-bin-list.rb
# This script generates the bin-list.zeek script.
# It uses data from: https://github.com/iannuttall/binlist-data

require 'csv'
require 'set'

puts "@load ./bank"
puts "module CreditCardExposure;"
puts
puts "event zeek_init()"
puts "	{"
puts "	if ( CreditCardExposure::load_bin_list )"
puts "		{"

i = 0
bins=Set.new
CSV.foreach("binlist-data.csv") do |row|
	i+=1
	next if i == 1
	next if row[0] !~ /^[0-9]+$/
	bin = row[0]
	next if bins.include?(bin)
	bins << bin

	brand = row[1].to_s.gsub(/\"/, "\\\"")
	type = row[2].to_s.gsub(/\"/, "\\\"")
	name = row[4].to_s.gsub(/\"/, "\\\"")
	country = row[5].to_s.gsub(/\"/, "\\\"")
	puts "		CreditCardExposure::bin_list[#{bin}] = Bank($brand=\"#{brand}\", $typ=\"#{type}\", $name=\"#{name}\", $country=\"#{country}\");"
end

puts "		}"
puts "	}"

File: auxil/process-bin-list.py
#!/usr/bin/env python3
# This script generates the bin-list.zeek script.
# It uses data from: https://github.com/iannuttall/binlist-data

import csv
from itertools import count
from collections import namedtuple
import operator

Bank = namedtuple("Bank", "brand type name country")
Item = namedtuple("Item", "bank bin")

HEADER = """
@load ./bank
module CreditCardExposure;

event zeek_init()
	{
	if ( !CreditCardExposure::load_bin_list )
		return;
"""

FOOTER = """
	}
"""


def clean(s):
    return s.replace('"', '\\"')


def read_file():
    with open("binlist-data.csv") as f:
        yield from csv.DictReader(f)


def load_data():
    bins = set()

    rows = []
    for row in read_file():
        bin = row["bin"]
        if not bin.isdigit():
            continue
        if bin in bins:
            continue
        bins.add(bin)

        brand = clean(row["brand"])
        type = clean(row["type"])
        name = clean(row["issuer"])
        country = clean(row["country"])

        bank = Bank(brand, type, name, country)
        rows.append(Item(bin=bin, bank=bank))

    rows.sort()
    return rows


def main():
    rows = load_data()

    print(HEADER)
    print("\tlocal b = CreditCardExposure::bin_list;")
    print("\tlocal l: Bank;")

    last = None
    for r in rows:
        b = r.bank
        if last and b == last.bank:
            print(f"\tb[{r.bin}]=l;")
        else:
            print(f'\tb[{r.bin}]=l=Bank($brand="{b.brand}", $typ="{b.type}", $name="{b.name}", $country="{b.country}");')
            last = r

    print(FOOTER)


if __name__ == "__main__":
    main()
